netty/src/main/java/io/grpc/netty/NettyChannelBuilder.java

```diff
@@ -290,16 +308,14 @@ static ProtocolNegotiator createProtocolNegotiator(
 
     private boolean closed;
 
-    private NettyTransportFactory(Class<? extends Channel> channelType,
-                                  NegotiationType negotiationType,
-                                  ProtocolNegotiator protocolNegotiator,
-                                  SslContext sslContext,
-                                  EventLoopGroup group,
-                                  int flowControlWindow,
-                                  int maxMessageSize,
-                                  int maxHeaderListSize) {
+    private NettyTransportFactory(
+        Class<? extends Channel> channelType, Map<ChannelOption<?>, Object> channelOptions,
+        NegotiationType negotiationType, ProtocolNegotiator protocolNegotiator,
+        SslContext sslContext, EventLoopGroup group, int flowControlWindow, int maxMessageSize,
+        int maxHeaderListSize) {
       this.channelType = channelType;
       this.negotiationType = negotiationType;
+      this.channelOptions = channelOptions;
```

Comment:
The copy should be made here, not in `NettyClientTransport`. `NettyClientTransport` is created _much_ later, each time we want to create a new connection. So that gives plenty of time for the map to mutate before the defensive copy is made. At that point, the builder is unlikely to modify the map. Copying here also means we only need one defensive copy for all the client transports created (they can share).